[Profiler] Make `LibKinetoClient::stop()` directly call `ProfilerStateBase::pop`

robieta opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84276
* #84275
* #84274
* #84273
* __->__ #83965

It has been discussed earlier in the stack at length, but if profiler fails after it pops the profiler state but before stopping Kineto then the next profiler call will see `LibKinetoClient::stop()` try to clean up the prior run (which it still thinks is active) by calling `disableProfiler()`. (Which fails because there is not an active profiler.)

This PR addresses the issue rather bluntly by simply rug pulling any active profiler from `LibKinetoClient::stop()`. I'm not particularly fond of this solution and we should refine the semantics in the future, but for now it has the desired effect of returning to a clean state. Earlier PRs in this stack cleaned up some of the lifetime management such that objects being destroyed triggers appropriate cleanup. As a result it is no longer catastrophic to simply pop the profiler state and let the destructor chain clean up.

Differential Revision: [D38958237](https://our.internmc.facebook.com/intern/diff/D38958237/)

Commit:
[Profiler] Make `LibKinetoClient::stop()` directly call `ProfilerStateBase::pop`

It has been discussed earlier in the stack at length, but if profiler fails after it pops the profiler state but before stopping Kineto then the next profiler call will see `LibKinetoClient::stop()` try to clean up the prior run (which it still thinks is active) by calling `disableProfiler()`. (Which fails because there is not an active profiler.)

This PR addresses the issue rather bluntly by simply rug pulling any active profiler from `LibKinetoClient::stop()`. I'm not particularly fond of this solution and we should refine the semantics in the future, but for now it has the desired effect of returning to a clean state. Earlier PRs in this stack cleaned up some of the lifetime management such that objects being destroyed triggers appropriate cleanup. As a result it is no longer catastrophic to simply pop the profiler state and let the destructor chain clean up.

Differential Revision: [D38958237](https://our.internmc.facebook.com/intern/diff/D38958237/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Make `LibKinetoClient::stop()` directly call `ProfilerStateBase::pop`"


It has been discussed earlier in the stack at length, but if profiler fails after it pops the profiler state but before stopping Kineto then the next profiler call will see `LibKinetoClient::stop()` try to clean up the prior run (which it still thinks is active) by calling `disableProfiler()`. (Which fails because there is not an active profiler.)

This PR addresses the issue rather bluntly by simply rug pulling any active profiler from `LibKinetoClient::stop()`. I'm not particularly fond of this solution and we should refine the semantics in the future, but for now it has the desired effect of returning to a clean state. Earlier PRs in this stack cleaned up some of the lifetime management such that objects being destroyed triggers appropriate cleanup. As a result it is no longer catastrophic to simply pop the profiler state and let the destructor chain clean up.

Differential Revision: [D38958237](https://our.internmc.facebook.com/intern/diff/D38958237/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Make `LibKinetoClient::stop()` directly call `ProfilerStateBase::pop`"


It has been discussed earlier in the stack at length, but if profiler fails after it pops the profiler state but before stopping Kineto then the next profiler call will see `LibKinetoClient::stop()` try to clean up the prior run (which it still thinks is active) by calling `disableProfiler()`. (Which fails because there is not an active profiler.)

This PR addresses the issue rather bluntly by simply rug pulling any active profiler from `LibKinetoClient::stop()`. I'm not particularly fond of this solution and we should refine the semantics in the future, but for now it has the desired effect of returning to a clean state. Earlier PRs in this stack cleaned up some of the lifetime management such that objects being destroyed triggers appropriate cleanup. As a result it is no longer catastrophic to simply pop the profiler state and let the destructor chain clean up.

Differential Revision: [D38958237](https://our.internmc.facebook.com/intern/diff/D38958237/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Make `LibKinetoClient::stop()` directly call `ProfilerStateBase::pop`"


It has been discussed earlier in the stack at length, but if profiler fails after it pops the profiler state but before stopping Kineto then the next profiler call will see `LibKinetoClient::stop()` try to clean up the prior run (which it still thinks is active) by calling `disableProfiler()`. (Which fails because there is not an active profiler.)

This PR addresses the issue rather bluntly by simply rug pulling any active profiler from `LibKinetoClient::stop()`. I'm not particularly fond of this solution and we should refine the semantics in the future, but for now it has the desired effect of returning to a clean state. Earlier PRs in this stack cleaned up some of the lifetime management such that objects being destroyed triggers appropriate cleanup. As a result it is no longer catastrophic to simply pop the profiler state and let the destructor chain clean up.

Differential Revision: [D38958237](https://our.internmc.facebook.com/intern/diff/D38958237/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Make `LibKinetoClient::stop()` directly call `ProfilerStateBase::pop`"


It has been discussed earlier in the stack at length, but if profiler fails after it pops the profiler state but before stopping Kineto then the next profiler call will see `LibKinetoClient::stop()` try to clean up the prior run (which it still thinks is active) by calling `disableProfiler()`. (Which fails because there is not an active profiler.)

This PR addresses the issue rather bluntly by simply rug pulling any active profiler from `LibKinetoClient::stop()`. I'm not particularly fond of this solution and we should refine the semantics in the future, but for now it has the desired effect of returning to a clean state. Earlier PRs in this stack cleaned up some of the lifetime management such that objects being destroyed triggers appropriate cleanup. As a result it is no longer catastrophic to simply pop the profiler state and let the destructor chain clean up.

Differential Revision: [D38958237](https://our.internmc.facebook.com/intern/diff/D38958237/)

[ghstack-poisoned]

slgong-fb commented on torch/csrc/profiler/kineto_client_interface.cpp:
```diff
@@ -45,7 +47,10 @@ class LibKinetoClient : public libkineto::ClientInterface {
 
   void stop() override {
-    (void)disableProfiler();
+    if (ProfilerStateBase::pop(/*global=*/true) ||
```
`ProfilerStateBase::pop()` would not be enough here? Or, would like to show this if statement cover both cases?
```
  static std::shared_ptr<ProfilerStateBase> pop() {
    auto out = pop(/*global=*/true);
    return out ? out : pop(/*global=*/false);
  }
```

Commit:
Update on "[Profiler] Make `LibKinetoClient::stop()` directly call `ProfilerStateBase::pop`"


It has been discussed earlier in the stack at length, but if profiler fails after it pops the profiler state but before stopping Kineto then the next profiler call will see `LibKinetoClient::stop()` try to clean up the prior run (which it still thinks is active) by calling `disableProfiler()`. (Which fails because there is not an active profiler.)

This PR addresses the issue rather bluntly by simply rug pulling any active profiler from `LibKinetoClient::stop()`. I'm not particularly fond of this solution and we should refine the semantics in the future, but for now it has the desired effect of returning to a clean state. Earlier PRs in this stack cleaned up some of the lifetime management such that objects being destroyed triggers appropriate cleanup. As a result it is no longer catastrophic to simply pop the profiler state and let the destructor chain clean up.

Differential Revision: [D38958237](https://our.internmc.facebook.com/intern/diff/D38958237/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Make `LibKinetoClient::stop()` directly call `ProfilerStateBase::pop`"


It has been discussed earlier in the stack at length, but if profiler fails after it pops the profiler state but before stopping Kineto then the next profiler call will see `LibKinetoClient::stop()` try to clean up the prior run (which it still thinks is active) by calling `disableProfiler()`. (Which fails because there is not an active profiler.)

This PR addresses the issue rather bluntly by simply rug pulling any active profiler from `LibKinetoClient::stop()`. I'm not particularly fond of this solution and we should refine the semantics in the future, but for now it has the desired effect of returning to a clean state. Earlier PRs in this stack cleaned up some of the lifetime management such that objects being destroyed triggers appropriate cleanup. As a result it is no longer catastrophic to simply pop the profiler state and let the destructor chain clean up.

Differential Revision: [D38958237](https://our.internmc.facebook.com/intern/diff/D38958237/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Make `LibKinetoClient::stop()` directly call `ProfilerStateBase::pop`"


It has been discussed earlier in the stack at length, but if profiler fails after it pops the profiler state but before stopping Kineto then the next profiler call will see `LibKinetoClient::stop()` try to clean up the prior run (which it still thinks is active) by calling `disableProfiler()`. (Which fails because there is not an active profiler.)

This PR addresses the issue rather bluntly by simply rug pulling any active profiler from `LibKinetoClient::stop()`. I'm not particularly fond of this solution and we should refine the semantics in the future, but for now it has the desired effect of returning to a clean state. Earlier PRs in this stack cleaned up some of the lifetime management such that objects being destroyed triggers appropriate cleanup. As a result it is no longer catastrophic to simply pop the profiler state and let the destructor chain clean up.

Differential Revision: [D38958237](https://our.internmc.facebook.com/intern/diff/D38958237/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Make `LibKinetoClient::stop()` directly call `ProfilerStateBase::pop`"


It has been discussed earlier in the stack at length, but if profiler fails after it pops the profiler state but before stopping Kineto then the next profiler call will see `LibKinetoClient::stop()` try to clean up the prior run (which it still thinks is active) by calling `disableProfiler()`. (Which fails because there is not an active profiler.)

This PR addresses the issue rather bluntly by simply rug pulling any active profiler from `LibKinetoClient::stop()`. I'm not particularly fond of this solution and we should refine the semantics in the future, but for now it has the desired effect of returning to a clean state. Earlier PRs in this stack cleaned up some of the lifetime management such that objects being destroyed triggers appropriate cleanup. As a result it is no longer catastrophic to simply pop the profiler state and let the destructor chain clean up.

Differential Revision: [D38958237](https://our.internmc.facebook.com/intern/diff/D38958237/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Make `LibKinetoClient::stop()` directly call `ProfilerStateBase::pop`"


It has been discussed earlier in the stack at length, but if profiler fails after it pops the profiler state but before stopping Kineto then the next profiler call will see `LibKinetoClient::stop()` try to clean up the prior run (which it still thinks is active) by calling `disableProfiler()`. (Which fails because there is not an active profiler.)

This PR addresses the issue rather bluntly by simply rug pulling any active profiler from `LibKinetoClient::stop()`. I'm not particularly fond of this solution and we should refine the semantics in the future, but for now it has the desired effect of returning to a clean state. Earlier PRs in this stack cleaned up some of the lifetime management such that objects being destroyed triggers appropriate cleanup. As a result it is no longer catastrophic to simply pop the profiler state and let the destructor chain clean up.

Differential Revision: [D38958237](https://our.internmc.facebook.com/intern/diff/D38958237/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Make `LibKinetoClient::stop()` directly call `ProfilerStateBase::pop`"


It has been discussed earlier in the stack at length, but if profiler fails after it pops the profiler state but before stopping Kineto then the next profiler call will see `LibKinetoClient::stop()` try to clean up the prior run (which it still thinks is active) by calling `disableProfiler()`. (Which fails because there is not an active profiler.)

This PR addresses the issue rather bluntly by simply rug pulling any active profiler from `LibKinetoClient::stop()`. I'm not particularly fond of this solution and we should refine the semantics in the future, but for now it has the desired effect of returning to a clean state. Earlier PRs in this stack cleaned up some of the lifetime management such that objects being destroyed triggers appropriate cleanup. As a result it is no longer catastrophic to simply pop the profiler state and let the destructor chain clean up.

Differential Revision: [D38958237](https://our.internmc.facebook.com/intern/diff/D38958237/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Make `LibKinetoClient::stop()` directly call `ProfilerStateBase::pop`"


It has been discussed earlier in the stack at length, but if profiler fails after it pops the profiler state but before stopping Kineto then the next profiler call will see `LibKinetoClient::stop()` try to clean up the prior run (which it still thinks is active) by calling `disableProfiler()`. (Which fails because there is not an active profiler.)

This PR addresses the issue rather bluntly by simply rug pulling any active profiler from `LibKinetoClient::stop()`. I'm not particularly fond of this solution and we should refine the semantics in the future, but for now it has the desired effect of returning to a clean state. Earlier PRs in this stack cleaned up some of the lifetime management such that objects being destroyed triggers appropriate cleanup. As a result it is no longer catastrophic to simply pop the profiler state and let the destructor chain clean up.

Differential Revision: [D38958237](https://our.internmc.facebook.com/intern/diff/D38958237/)

[ghstack-poisoned]

robieta commented:
@pytorchbot merge

robieta commented:
@pytorchbot revert -m "broke internal on-demand tracing: S296407"

robieta commented:
@pytorchbot revert -m "broke internal on-demand tracing: S296407" -c nosignal
